Add support for raw mode

FriedlandAaron opened the pull request:
This PR is an attempt to add support for raw mode, as discussed here #79.

Quick overview of changes:

- added --raw-mode and -r flags to both node and rust APIs
- implemented raw mode in both node and rust APIs
- updated relevant tests to reflect the flag addition
- updated relevant tests to reflect the implementation
- added new tests for the implementation

Hope everything looks good!

FriedlandAaron committed:
implement rust api raw mode flag

FriedlandAaron committed:
add doc tests for raw mode

FriedlandAaron committed:
update doc test to test all long-form flags

FriedlandAaron committed:
implement node api raw mode flag

FriedlandAaron committed:
add and update tests for raw mode support

FriedlandAaron committed:
fix lint warnings

dominikwilkowski commented on nodejs/src/Render.js:
```diff
@@ -242,6 +242,18 @@ const Render = (input, SETTINGS = {}, debug = DEBUG.enabled, debuglevel = DEBUG.
 			`</div>`;
 	}
 
+	// handle raw mode
+	if (OPTIONS.env === 'node') {
+		if (OPTIONS.rawMode === true) {
+			const re = /\n/g;
+			const newline = '\r\n';
+			write = write.replace(re, newline);
+			output = output.map((x) => {
+				return x.replace(re, newline);
+			});
+		}
+	}
```
Instead of replacing after-the-fact, maybe add a variable at the top that defaults to `\n` and is changed with the option later to `\r\n`?

FriedlandAaron replied:
Can do, but it would require a bit more refactoring to the logic. I would want the OK from you if you think we should make those changes.
See #79 for more details.

dominikwilkowski commented on rust/src/render.rs:
```diff
@@ -215,6 +229,16 @@ pub fn render(options: Options) -> RenderedString {
 		d("render() formatted for Env::Browser", 1, Dt::Log, &options, &mut std::io::stdout());
 	}
 
+	// Handle raw mode only in Cli env
+	if options.env == Env::Cli && options.raw_mode {
+		for v in output.iter_mut() {
+			*v = v.replace('\n', "\r\n");
+		}
+		text = text.replace('\n', "\r\n");
+	}
```
Same as above

FriedlandAaron replied:
Same as above :smile:

dominikwilkowski commented on rust/src/render.rs:
```diff
@@ -50,6 +50,20 @@ pub struct RenderedString {
 ///             " █▀█ ██▄ █▄▄ █▄▄ █▄█\n\n"
 ///         )
 ///     );
+///        let output = render(Options {
```
```suggestion
///
///     let output = render(Options {
```

FriedlandAaron replied:
Will do

dominikwilkowski commented on rust/src/render.rs:
```diff
@@ -50,6 +50,20 @@ pub struct RenderedString {
+///         raw_mode: true,
+///         ..Options::default()
+///     });
+///     assert_eq!(
```
```suggestion
///
///     assert_eq!(
```

FriedlandAaron replied:
Will do

dominikwilkowski commented on rust/src/config.rs:
```diff
@@ -495,4 +498,13 @@ pub const CLIOPTIONS: [CliOption; 16] = [
+		name: "--raw-mode",
+		shortcut: "-r",
+		fallback_shortcut: "",
+		description: "Use to enable proper newline rendering for 'raw mode' in terminals",
```
I'd probably add that we add `\r` to line breaks? What do you think? (changes here should be reflected in the node help as well)

FriedlandAaron replied:
Great idea! Will add that

dominikwilkowski commented on rust/src/args.rs:
```diff
@@ -36,6 +36,7 @@ use crate::debug::{d, Dt};
 /// options.version = true;
 /// options.debug = true;
 /// options.debug_level = 3;
+/// options.raw_mode = true;
```
Nit: I'd love the keep the debug flags at the bottom and version right above them?
I'd even move it further up to right below `env`. Seems appropriate?

FriedlandAaron replied:
Sure! I don't mind nit-picking at all. It's your project, after all :)

dominikwilkowski commented on nodejs/test/unit/Render.spec.js:
```diff
@@ -958,3 +980,189 @@ test(`Render - Returns false if font is not known`, () => {
+			' ╚═╝  ╚═╝    ╚═╝   \r\n\r\n'
+	);
+	expect(test.array).toEqual([
+		'\r\n\r\n ████████╗ ███████╗',
```
I'd move the `\r\n\r\n` to a new line just like the others so it's more readable.

FriedlandAaron replied:
Makes sense. I'll probably have to go over the other tests to see if I can apply this to other places

dominikwilkowski reviewed (changes requested):
Thank you so much for this.
This is probably the best PR I have gotten for this project in it's somewhat 20ish year streak in the open source :)

I did leave some comments here and there with some nits I'm also happy to do myself. The biggest one would be to use a variable to add new lines and change it after we parsed the options. Replace is super quick and wouldn't really add much performance issues but it's just cleaner coding. Happy to also do this myself but figured I chat to you first ;)

FriedlandAaron commented:
Hey, first of all, thanks for the kind words!!

I'll write my input into your comments, but in general, all seem like good ideas, and I don't mind doing them.

Regarding the newline variable:
I was considering just adding a function that received the rawmode flag and returned the appropriate newline char,  but I ended up going with the current solution because of the inconsistency I mentioned here #79. A variable would also work, like you suggest, but it would still mean slightly different changes in different locations. Not a huge deal, but I decided not to do it, in favor of changing as little of the implementation as possible. I'll go into detail in the issue conversation #79.

FriedlandAaron committed:
update help for raw mode flag

FriedlandAaron committed:
reposition raw mode arg

FriedlandAaron committed:
fix formatting

FriedlandAaron committed:
manual alignment of test text

dominikwilkowski commented on nodejs/src/index.js:
```diff
@@ -49,10 +49,11 @@ const Cli = (inputOptions = CLIOPTIONS, inputArgs = process.argv) => {
 			`    env: ${args.env},\n` +
-			`  }, ${args.debug}, ${args.debugLevel} );`,
+			`    rawMode: ${args['raw-mode']},\n` +
+			`  }, ${args.debug}, ${args['debug-level']} );`,
```
Wait was this a bug? Did it mention the wrong casing here?

FriedlandAaron replied:
Apparently. The args variable is a map created based off of CLIOPTIONS, so the keys are based on the cli flags, as opposed to the Options object used in other parts of the code.

FriedlandAaron committed:
change space to spaceless in nodejs implementation

FriedlandAaron commented:
As per our discussion in #79 regarding the spaceless option in node:

Changed the 'space' option to 'spaceless' in the nodejs implementation.
- updated the Options object
- updated the tests

FriedlandAaron committed:
handle line break at start of render function

FriedlandAaron commented:
As we discussed in #79, the implementation was changed to use a variable that is set based on the relevant flags.
Tests pass when I run them locally.

dominikwilkowski commented on rust/src/render.rs:
```diff
@@ -196,28 +203,28 @@ pub fn render(options: Options) -> RenderedString {
 		match options.align {
-			Align::Top => output.push(String::from("\n\n\n")),
-			Align::Bottom => output[0] = String::from("\n\n\n\n") + &output[0],
+			Align::Top => output.push(format!("{line_break}{line_break}{line_break}")),
```
I just had a look at `format!("{line_break}{line_break}{line_break}")` vs `line_break.repeat(3).to_string()` in terms of performance but it all sort of washes away because of the `.to_string()` call. I wonder if making `line_break` a `String` from the start (a single heap allocation) and then just use `.repeat` where you need it?
But then we don't have that many occurrences here that actually need this.
Never mind me... probably just getting distracted. _(I thought it was still worth sharing)_

FriedlandAaron replied:
I do agree it's pretty ugly, but I don't think there's a big difference here. The repeat option is cleaner, you could also store the amount of repeats as a variable and modify that really easily in the future if you wanted. Would you prefer that?

dominikwilkowski replied:
I don't think you need to do anything in this PR. Overall I need to redo how I handle strings in this library. I think it's wildly inefficient. When I wrote this rust port I was locked in a hotel room in Toronto because of covid for 10 days and may have also hallucinated things 😬 
Using the `format` macro should be kept to a monimum and there is probably a better way to deal with how we build those strings by using a single mutable reference. One day I have time for this :)

dominikwilkowski replied:
Fixing the tests is the only outstanding thing for this to land

FriedlandAaron replied:
Are there tests that are not passing? On my local machine, `cargo test` is passing all tests.

dominikwilkowski replied:
I'm talking about the CI tests. GitHub actions are failing because some of them might be outdated. I'll try to fix it and then we can merge that fix into this branch and merge to main.

dominikwilkowski reviewed (approved):
This looks great now!
Thank you

dominikwilkowski commented:
Looks like the action is failing due to old node versions? I have to experiment with this in another PR. Give me a day or two

dominikwilkowski committed:
Merge branch 'released' into feat/raw-mode-flag
